### ISSUE TITLE
请问一下 focal loss会有提升吗？yolov3中使用好像精度会下降

### ISSUE THREAD

**Issue by zhanghongsir** [AkitsukiM/YOLOv3-Pytorch-VOC#1]

**Comment by AkitsukiM** [comment]
VOC数据集上跑了2次实验，发现在损失函数中用BCE代替Focal Loss时mAP会下降约0.03（0.84 -> 0.81）
COCO数据集验证中，代码晚些时候上传……

**Comment by zhanghongsir** [comment]
> VOC数据集上跑了2次实验，发现在损失函数中用BCE代替Focal Loss时mAP会下降约0.03（0.84 -> 0.81）
> COCO数据集验证中，代码晚些时候上传……

我也进行了实验，结果确实是会下降2到3个点，但是我不明白原因。focal loss在这里是求置信度损失的，为什么不用在分类损失里呢（我在这里进行了实验，将分类的bcewithlogitsloss换成了focal loss精度下降了一个点）。还有就是我的数据集是单类的，我查阅资料说，数据集单类在yolov3中分类损失要为0。请问是这样的吗？

**Comment by AkitsukiM** [comment]
> > VOC数据集上跑了2次实验，发现在损失函数中用BCE代替Focal Loss时mAP会下降约0.03（0.84 -> 0.81）
> > COCO数据集验证中，代码晚些时候上传……
> 
> 我也进行了实验，结果确实是会下降2到3个点，但是我不明白原因。focal loss在这里是求置信度损失的，为什么不用在分类损失里呢（我在这里进行了实验，将分类的bcewithlogitsloss换成了focal loss精度下降了一个点）。还有就是我的数据集是单类的，我查阅资料说，数据集单类在yolov3中分类损失要为0。请问是这样的吗？

Focal Loss主要是为了解决样本数不均衡的问题，所以如果各类别样本数在统计上比较均衡就只需要用BCE。置信度用FL是因为一个图有上百个grid cell但只有几个正例（ground truth），剩下的都是反例
像YOLO这样的单阶段检测器在单类目标检测任务里就没有分类损失了，反例框的抑制由置信度即可完成。如果是双阶段检测器还是有分类损失的（正反例分类）

**Comment by zhanghongsir** [comment]
> > > VOC数据集上跑了2次实验，发现在损失函数中用BCE代替Focal Loss时mAP会下降约0.03（0.84 -> 0.81）
> > > COCO数据集验证中，代码晚些时候上传……
> > 
> > 
> > 我也进行了实验，结果确实是会下降2到3个点，但是我不明白原因。focal loss在这里是求置信度损失的，为什么不用在分类损失里呢（我在这里进行了实验，将分类的bcewithlogitsloss换成了focal loss精度下降了一个点）。还有就是我的数据集是单类的，我查阅资料说，数据集单类在yolov3中分类损失要为0。请问是这样的吗？
> 
> Focal Loss主要是为了解决样本数不均衡的问题，所以如果各类别样本数在统计上比较均衡就只需要用BCE。置信度用FL是因为一个图有上百个grid cell但只有几个正例（ground truth），剩下的都是反例
> 像YOLO这样的单阶段检测器在单类目标检测任务里就没有分类损失了，反例框的抑制由置信度即可完成。如果是双阶段检测器还是有分类损失的（正反例分类）

谢谢您的解答，但是我输出代码里的分类损失的时候为啥还是有值的一直在波动。。。我需要手动把它设置为0吗？（我确实也这样做了，但是好像没什么用）

**Comment by AkitsukiM** [comment]
> 谢谢您的解答，但是我输出代码里的分类损失的时候为啥还是有值的一直在波动。。。我需要手动把它设置为0吗？（我确实也这样做了，但是好像没什么用）

额，这个时候网络输出的通道数就应该是ANCHORS_PER_SCLAE * 5而不是ANCHORS_PER_SCLAE * (num_classes + 5)了
model、loss function和evaluator需要同时改（否则[5:]之类的index操作都会报错）
还有一种可能的办法是在cfg里把NUM_CLASSES改为1，把类别输出切片的值都置为20（趋于1的反sigmoid，弄一个稍大的数都可以），分类损失项直接删掉，应该就没问题了吧（其实我也没试过（笑）

**Comment by zhanghongsir** [comment]
> > 谢谢您的解答，但是我输出代码里的分类损失的时候为啥还是有值的一直在波动。。。我需要手动把它设置为0吗？（我确实也这样做了，但是好像没什么用）
> 
> 额，这个时候网络输出的通道数就应该是ANCHORS_PER_SCLAE * 5而不是ANCHORS_PER_SCLAE * (num_classes + 5)了
> model、loss function和evaluator需要同时改（否则[5:]之类的index操作都会报错）
> 还有一种可能的办法是在cfg里把NUM_CLASSES改为1，把类别输出切片的值都置为20（趋于1的反sigmoid，弄一个稍大的数都可以），分类损失项直接删掉，应该就没问题了吧（其实我也没试过（笑）

不是吧  我咋看别的代码都是ANCHORS_PER_SCLAE * (num_classes + 5)。。。。。要改这么多吗
